[Preprocessing] Add support for general ContractionOps and handle dynamic dimension using pad + expand.

raikonenfnu opened the pull request:
During LLM prefill phase, we do not know the shape/context length is not pre-determined. Compiling with pure dynamic shape would not give us best performance since we cannot utilize the HW intrinsics. 

This PR re-introduce PadToIntrinsic for ContractionOps and introduces Pad-Expand pipeline to support dynamic shape/dims. The general idea is we would pad the dynamic dimension until it is divisible by the intrinsic size, and then we would expand the shape by that intrinsic size to "materialize" it. 

This required us to do two things. The first is to modified existing code to be more OpFoldResult such that we can compute dynamic shapes. Secondly, we also introduce some code to keep track of the metadata of the dimensions in order for us to generate appropriate for padOp, expandOp, as well as a correct expanded-genericOp with the appropriate shapes, iteratorTypes, and dimensions.

raikonenfnu committed:
[Preprocessing] Handle/pad dynamic dimension using pad + expand.

raikonenfnu committed:
Update bazel + cmake

raikonenfnu committed:
Update comments for better explanation and fix typo.

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -30,26 +31,116 @@ namespace mlir::iree_compiler::Preprocessing {
-                            Value padSource, ArrayRef<int64_t> padding) {
-  auto sourceType = cast<RankedTensorType>(padSource.getType());
+                            Value padSource, ArrayRef<OpFoldResult> padding) {
+  auto sourceType = padSource.getType().cast<RankedTensorType>();
```
This way is being deprecated. What is there before is the way the casting needs to be done.

raikonenfnu replied:
done!

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -30,26 +31,116 @@ namespace mlir::iree_compiler::Preprocessing {
+expandMapsAndIterators(SmallVector<AffineMap> &expandedMaps,
+                       SmallVector<utils::IteratorType> &expandedIterators,
+                       SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  int expandOffset = 0;
```
Could this logic be made simpler with https://github.com/shark-infra/llvm-project/blob/c9866736ccbd55026af03c8e3ceee41a6186d63c/mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp#L32

raikonenfnu replied:
Hey Mahesh, I think with we can use `getReassociationIndicesForReshape` to get some metadata, and would be able to query the expanded dims from reassociation indices instead of using:
```c
      int64_t expandSrcDim = expandDim + expandOffset;
      int64_t expandDstDim = expandSrcDim + 1;
```

This would simplify the dim tracking and offset-dimension a bit. But this may add extra complexity in logic since reassociation indices is dimension from operand POV, while we need to also handle iteratorTypes from the linalgOp's POV and shuttling of srcShape and targetShape of the operands. Any thoughts?

MaheshRavishankar replied:
Ok, I am fine with leaving this as is.

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -198,15 +504,26 @@ struct PadToIntrinsicsPass final
-         llvm::make_early_inc_range(targetOps)) {
-      rewriter.setInsertionPoint(convOp);
-      padConvOp(rewriter, convOp, intrinsics);
+    for (auto linalgOp : llvm::make_early_inc_range(targetOps)) {
```
Ok, my bad here (probably dont know who committed the original sin), but better to walk the program and collect the operations to transform and then do the transformation (to avoid doing `make_early_inc_range`. Will also help prune `GenericOp` that are not contraction ops during the walk.

raikonenfnu replied:
Sorry Mahesh, I am not sure if I fully understand. Aren't we already doing a rootOp->walk? Should we update the rootOp->walk to only add `GenericOp` if `linalg::isaContractionOpInterface` is true and replace the `make_early_inc_range` with regular loop? 

Also is there anything wrong with `make_early_inc_range`? haha

kuhar replied:
early inc range allows you to iterate over (intrusive) link lists whose nodes get invalidated by the current iteration. We can use a regular for loop here, I missed this when cleaning up this code.

raikonenfnu replied:
I removed the make_early_inc_range, and only pushed in ops that matches our option into the worklist.

MaheshRavishankar reviewed (changes requested):
I think it might be better to provide some optionality here. Maybe have a enum pass option that allows you to specify if  the pass should pad contraction ops or convolution ops. It might not be beneficial for all use cases to pad convolution and contraction?

benvanik commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -198,15 +504,26 @@ struct PadToIntrinsicsPass final
-    rootOp->walk(
-        [&](linalg::Conv2DNhwcHwcfOp convOp) { targetOps.push_back(convOp); });
+    SmallVector<linalg::LinalgOp> targetOps;
+    rootOp->walk([&](linalg::LinalgOp linalgOp) {
```
note that this is unsafe - root walks are not allowed outside of codegen as this will walk into any executables that already exist and mess with them - instead you should loop over FunctionOpInterface ops and walk those

(if a user has partially compiled the program, used custom ops/dispatches, tk, etc there will be executables in the input)

raikonenfnu replied:
Hey Ben, I modified the code to be FunctionOpInterface so it will just walk on funcOpInterface operations. I also added the more "local" look up and exit if the number of targetAttrs found is none-1.

raikonenfnu committed:
Address comments

1. New cast API
2. Use funcOpInterface for safety
3. Localize lookup for executable targets.
4. Add target pad types as an option (between conv, contraction, and
   all)

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -30,35 +32,151 @@ namespace mlir::iree_compiler::Preprocessing {
+                       SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  int expandOffset = 0;
+  SmallVector<int64_t> dimsToExpandVec = llvm::to_vector(
+      llvm::map_range(dimsToExpand, [](auto &dim) { return dim.first; }));
```
```suggestion
  auto dimsToExpandVec = llvm::to_vector_of<int64_t>(
      llvm::make_first_range(dimsToExpand));
```

raikonenfnu replied:
woahh, didn't know this exist, that's cool!

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -30,35 +32,151 @@ namespace mlir::iree_compiler::Preprocessing {
+                             expandedIterators[expandSrcDim]);
+    // Updating map of each operand to handle newly expanded/dst dim
+    // based on the location of it's expand source dim.
+    for (int operandIdx = 0; operandIdx < expandedMaps.size(); operandIdx++) {
```
nit: use range for

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+  SmallVector<std::array<OpFoldResult, 3>> mnkPaddingCandidates;
+  // Compute Pad and Expand metadata for linalgOp's M, N, K dimensions. Multiple
+  // candidates of the metadata will be formed based on different intrinsics.
+  for (auto &intrinsic : intrinsics) {
```
nit: write the type when it's not obvious by looking at the RHS

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+      std::optional<unsigned> maybeDim = operandMap.getResultPosition(
+          getAffineDimExpr(targetDim, operandMap.getContext()));
+      if (maybeDim)
+        return std::make_pair(operand->get(), maybeDim.value());
```
nit: you can construct pair directly, `make_pair` is mostly a c++98 thing now :P

```suggestion
        return std::pair{operand->get(), maybeDim.value()};
```

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+  // Util to get operand and dim;
+  auto getSrcOperandAndDim =
+      [&](int64_t targetDim) -> std::optional<std::pair<Value, int64_t>> {
+    for (auto operand : linalgOp.getDpsInputOperands()) {
```
nit: write the type when it's not obvious by looking at the RHS

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+  // extract slice.
+  auto resultType = cast<RankedTensorType>(linalgOp->getResult(0).getType());
+  auto resultShape = resultType.getShape();
+  auto resultRank = resultType.getRank();
```
also here

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+    }
+  }
+
+  // extract slice.
```
```suggestion
  // Extract slice.
```

raikonenfnu replied:
done!

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +290,271 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+        .Case<linalg::BatchMatmulOp, linalg::MatmulOp,
+              linalg::MatmulTransposeBOp, linalg::GenericOp>(
+            [&](auto matmulOp) { padContractionLikeOp(rewriter, linalgOp); })
+        .Default([&](Operation *op) {});
```
unused argument `op`. also, shouldn't we assert this doesn't happen?

raikonenfnu replied:
Yeah, this shouldn't happen. I added an assert. :)

kuhar reviewed (approved):
Looks good, just a few local nits.
I haven't parsed through all the logic, but the tests look solid to me, so I believe it works.

raikonenfnu committed:
NIT and suggestions by Jakub.

kuhar commented on compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp:
```diff
@@ -172,45 +289,275 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+        .Default([&](Operation *op) {
+          // Should not happen since we only add ops we know into the
+          // worklist.
+          assert(false && "Unsupported op to pad.");
```
You know what, I think this would actually be more straightforward if we had 2 separate vectors instead of mixing all of these in a single one. The type switch and the assertion would go away

MaheshRavishankar replied:
Yeah, two separate vectors is what I was looking for as well.

raikonenfnu replied:
That makes sense, I think Mahesh recommended this as well earlier haha

raikonenfnu replied:
I just saw @MaheshRavishankar new reply on this after posting this comment, but yeah, changed/done. :)

MaheshRavishankar reviewed (approved):
Thanks! This looks good to me. Left one last comment.

raikonenfnu committed:
two worklist.

raikonenfnu committed:
NIT: reorder the conv and contract padding for more consistency.

kuhar reviewed (approved):
LGTM
